File: lib/custom_element.dart
import 'package:js/js.dart';
import 'dart:html';

@JS('defineCustomElement')
external void _defineCustomElement(
  String tag,
  List<String> observedAttributes,
  List<Function> Function(HtmlElement) getDartCustomElementHooks,
);

class CustomElement {
  final HtmlElement element;

  CustomElement(this.element);

  onConnected() {}
  onDisconnected() {}
  onAdopted() {}
  onAttributeChanged(String attrName, dynamic oldVal, dynamic newVal) {}
}

void defineCustomElement(
  String tag,
  List<String> observedAttributes,
  CustomElement Function(HtmlElement) creator,
) {
  _defineCustomElement(
    tag,
    observedAttributes,
    allowInterop((element) {
      final customElement = creator(element);
      return [
        allowInterop(customElement.onConnected),
        allowInterop(customElement.onDisconnected),
        allowInterop(customElement.onAdopted),
        allowInterop(customElement.onAttributeChanged),
      ];
    }),
  );
}

File: web/custom_element.js
function defineCustomElement(
  tag,
  observedAttributes,
  getDartCustomElementHooks
) {
  class CustomElement extends HTMLElement {
    static observedAttributes = observedAttributes;

    constructor() {
      super();
      const dartCustomElementHooks = getDartCustomElementHooks(this);
      this.onConnected = dartCustomElementHooks[0];
      this.onDisconnected = dartCustomElementHooks[1];
      this.onAdopted = dartCustomElementHooks[2];
      this.onAttributeChanged = dartCustomElementHooks[3];
    }

    connectedCallback() {
      this.onConnected();
    }

    disconnectedCallback() {
      this.onDisconnected();
    }

    adoptedCallback() {
      this.onAdopted();
    }

    attributeChangedCallback(attrName, oldVal, newVal, whatIsThis) {
      this.onAttributeChanged(attrName, oldVal, newVal);
    }
  }

  customElements.define(tag, CustomElement);
}
